src/Microsoft.NetCore.Analyzers/Core/Security/DoNotCallDangerousMethodsInDeserialization.cs

```diff
@@ -0,0 +1,219 @@
+                {
+                    var compilation = compilationStartAnalysisContext.Compilation;
+                    var attributeTypeSymbols = ImmutableArray.Create(
+                        WellKnownTypes.OnDeserializingAttribute(compilation),
+                        WellKnownTypes.OnDeserializedAttribute(compilation));
+                    var serializableAttributeTypeSymbol = WellKnownTypes.SerializableAttribute(compilation);
+                    var streamingContextTypeSymbol = WellKnownTypes.StreamingContext(compilation);
+
+                    if (serializableAttributeTypeSymbol == null || streamingContextTypeSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var builder = ImmutableArray.CreateBuilder<IMethodSymbol>();
+                    var systemIOFileTypeSymbol = WellKnownTypes.SystemIOFile(compilation);
+
+                    if (systemIOFileTypeSymbol != null)
+                    {
+                        foreach (var targetMethodName in SystemIOFileMethodMetadataNames)
+                        {
+                            builder.AddRange(systemIOFileTypeSymbol.GetMembers(targetMethodName).Cast<IMethodSymbol>());
+                        }
+                    }
+
+                    var systemReflectionAssemblyTypeSymbol = WellKnownTypes.SystemReflectionAssembly(compilation);
+
+                    if (systemReflectionAssemblyTypeSymbol != null)
+                    {
+                        foreach (var targetMethodName in SystemReflectionAssemblyMethodMetadataNames)
+                        {
+                            builder.AddRange(systemReflectionAssemblyTypeSymbol.GetMembers(targetMethodName).Cast<IMethodSymbol>());
+                        }
+                    }
+
+                    var dangerousMethodSymbols = builder.ToImmutable();
+
+                    if (dangerousMethodSymbols.Length == 0)
+                    {
+                        return;
+                    }
+
+                    compilationStartAnalysisContext.RegisterSymbolAction(
+                        (SymbolAnalysisContext symbolAnalysisContext) =>
+                        {
+                            var classSymbol = symbolAnalysisContext.Symbol as INamedTypeSymbol;
+
+                            if (!classSymbol.GetAttributes().Any(s => s.AttributeClass.Equals(serializableAttributeTypeSymbol)))
+                            {
+                                return;
+                            }
+                            
+                            foreach (var member in classSymbol.GetMembers())
+                            {
+                                if (member.Kind == SymbolKind.Method)
+                                {
+                                    var methodSymbol = member as IMethodSymbol;
+
+                                    if (methodSymbol.GetAttributes().Any(s => attributeTypeSymbols.Contains(s.AttributeClass))
```

Comment:
The query also looks at `Dispose()` and `Dispose(bool)` methods, as well as finalizers.  Want to look at those as well?

Reply:
We had a discussion about that. If we look at Dispose() and Finalizers, it will impact those classes which don't support serialization.

Reply:
Right, but if the class has `[Serializable]` on it, then it would be worth looking at.